### DISCUSSION TITLE
Is there an easy way to propagate `VCPKG_TARGET_TRIPLET` to a project using Manifest?

### DISCUSSION THREAD

**Discussion by bwrsandman** [microsoft/vcpkg#32057] [Q&A] [answered]
Given this basic manifest project, is there a way to easily configure it for cross compiling? Seems like `VCPKG_TARGET_TRIPLET` only configures the dependencies to be of that triplet, it would stand to reason that this configuration could be done for the actual project by leveraging the existing cmake code.

Ideally it could look like this (on a x64-linux host)
```bash
cmake -S. -Bcmake-build -DVCPKG_TARGET_TRIPLET=arm64-linux -DCMAKE_TOOLCHAIN_FILE=vcpkg/scripts/buildsystems/vcpkg.cmake
cmake --build cmake-build
```

## What I've tried

Currently what happens is that the packages are built using the `VCPKG_TARGET_TRIPLET` but the actual cmake project will compile to the host triplet and fail when trying to link. This happens if I try emscriptem, android, x86 and arm on windows, mac and linux hosts.

It is great for compiling the packages correctly but no configuration is passed to my project and it seems like I'm on my own to set the proper toolchains and settings.

What happens when I try is either something like this for arm64-linux
```
/usr/bin/ld: vcpkg_installed/arm64-linux/debug/lib/libspdlogd.a: error adding symbols: file in wrong format
$ file cmake-build/CMakeFiles/vcpkg-cross.dir/main.cpp.o # main.cpp is compiled in 64 bits architecture
cmake-build/CMakeFiles/vcpkg-cross.dir/main.cpp.o: ELF 64-bit LSB relocatable, x86-64, version 1 (SYSV), not stripped
```
Or this for x86-linux
```
CMake Error at vcpkg/scripts/buildsystems/vcpkg.cmake:853 (_find_package):
  Could not find a configuration file for package "spdlog" that is compatible
  with requested version "".

  The following configuration files were considered but not accepted:

    vcpkg-cross/cmake-build/vcpkg_installed/x86-linux/share/spdlog/spdlogConfig.cmake, version: 1.11.0 (32bit)

Call Stack (most recent call first):
  CMakeLists.txt:3 (find_package)

```

A similar question is asked in #26846 but the solution of using `VCPKG_HOST_TRIPLET` doesn't solve the issue.
1. It causes `host` dependencies to be cross compiled which is not useful.
2. The main issue of the project object files being compiled in the wrong triplet is still present.

### `vcpkg.json`
```json
{
    "name": "vcpkg-cross",
    "version-string": "0.0.1-dev",
    "dependencies": ["spdlog"]
}
```
### `CMakeLists.txt`
```cmake
cmake_minimum_required(VERSION 3.20)
project(vcpkg-cross)
find_package(spdlog REQUIRED)
add_executable(vcpkg-cross main.cpp)
target_link_libraries(vcpkg-cross PRIVATE spdlog::spdlog)
```
### `main.cpp`
```c++
#include <spdlog/spdlog.h>
int main(int argc, char* argv[])
{
    spdlog::info("Hello World!");
    return 0;
}
```

**Comment by autoantwort** [comment]
You can pass the target triplet via the command line `-DVCPKG_TARGET_TRIPLET=x64-windows` for example

**Reply by bwrsandman** [reply]
The issue is if the target is another architecture like `arm64` or even `x86`, the executable gets compiled to the host architecture (e.g. `x64`) and fails when linking with the packages which are (e.g. `arm64`).

**Comment by JackBoosY** [comment]
In manifest mode, you should pass the extra install options to cmake using `-DVCPKG_INSTALL_OPTIONS=<OPTIONS>`.

Also, to avoid triggering the cross build, you should set `VCPKG_HOST_TRIPLET` to the same value `x64-linux/arm64-linux` as `VCPKG_TARGET_TRIPLET`.

**Reply by bwrsandman** [reply]
This doesn't really answer my question. I am trying to cross build and I addressed `VCPKG_HOST_TRIPLET`

> A similar question is asked in https://github.com/microsoft/vcpkg/discussions/26846 but the solution of using VCPKG_HOST_TRIPLET doesn't solve the issue.
> 1.    It causes host dependencies to be cross compiled which is not useful.
> 2.   The main issue of the project object files being compiled in the wrong triplet is still present.

Host dependencies such as shader compilers should be compiled to the host triplet, that's behavior I would like to keep.

**Comment by dg0yt** [comment, marked as answer]
IIUC the question isn't really related to manifest mode, but to configuring the user project from `VCPKG_TARGET_TRIPLET` when using the vcpkg toolchain.

This is why it doesn't work at the moment for user projects:
 - The compiler setup is to be done by the CMake toolchain.
   With `vcpkg.cmake`, this is delegated to chainloaded toolchain files, `VCPKG_CHAINLOAD_TOOLCHAIN_FILE`.
 - `vcpkg.cmake` doesn't initialize `VCPKG_CHAINLOAD_TOOLCHAIN_FILE`, so the default CMake initialization takes place if the user doesn't set that variable.
   For port builds, it is initialized from the cmake maintainer functions.
- The specific vcpkg files used as chainloaded toolchain files, e.g. `scripts/toolchains/linux.cmake` rely on variables from the triplet file itself, e.g. `VCPKG_TARGET_ARCHITECTURE`. 
  However, the triplet files are never loaded in project mode.
  For port builds, selected variables are forwarded by the cmake maintainer functions.

So there is simply no transformation of `VCPKG_TARGET_TRIPLET` into CMake toolchain initialization for the user project, despite the capability to do that for ports :disappointed: 
What would be needed is to detect the user-provided value and ensure the same initialization as applied to ports if not overridden in other ways.

**Reply by bwrsandman** [reply]
Thank you for the explanation.

One thing I noticed is that there there is a lot of useful code in vcpkg.cmake for detecting platforms. I wonder if some of it could be used to reconstruct the triplets in project mode.
